keps/sig-cli/2551-return-code-normalization/README.md

```diff
@@ -0,0 +1,637 @@
+| 213   | An underlying service was unavailable                                                                 |
+| 214   | API internal error                                                                                    |
+| 215   | Too many requests                                                                                     |
+| 216   | Request entity too large                                                                              |
+| 217   | Unexpected response from API                                                                          |
+| 255   | An external command returned an exit code equal to or greater than 201, which is reserved for kubectl |
+
+* The reserved exit codes are documented [here](https://tldp.org/LDP/abs/html/exitcodes.html) and should be used carefully in a way to not generate conflict with existing scripts.
+* With consideration of exit codes from exec and other external commands (i.e. diff), starting kubectl codes at 201 would allow the original exit codes to pass through most of the time
+* Starting kubectl error exit codes at 201 allows them to be more distinct and reduce confusion as to the origin of the error, whether kubectl itself or an external command
+* The default kubectl error exit code would change from 1 to 201, with the hope that kubectl would never return the default error exit code, instead using a more appropriate code
+
+### Changing error checker functions
+All of the steps above uses [cmdutil.CheckErr](https://github.com/kubernetes/kubernetes/blob/2a26f276a8c8c13b2f45927ee5ece2063950dd1d/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L114) 
+function to delegate the error validation. This function might be slight changed 
+(without changing its signature) to instead of use the default `fatalErrHandler` 
+function, use some more specific function that might verify the error against a matrix of possible errors
+and exit with the right code.
+
+The function [CheckDiffErr](https://github.com/kubernetes/kubernetes/blob/2a26f276a8c8c13b2f45927ee5ece2063950dd1d/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L124)
+can be used as an example of how this can be implemented.
+
+The pro of using this design approach is that almost no change will be needed in commands, as they already use the 
+CheckErr function.
+
+The con of using this design approach is that the Error Validation functions are pretty hard to change/evolve
+in the manner they are designed, calling another checkErr inner function.
+
+For the rollout of this approach, a new non-exported error checker function needs to be developed and make the 
+call to the old checkErr or the new function according to the value of the Environment Variable described in
+[Risks and Mitigations](#risks-and-mitigations) 
+
+### Creating new error parser functions
+
+Another design solution is to create helper functions for each steps:
```

Comment:
I was thinking about it earlier this week and I think that implementing a set of new error codes inside kubectl where each command could use them and then having `CheckErr` return new error codes with that env of yours and just -1 in the backwards compatible way as it does today. Which is close with what you're describing here.

Reply:
sorry @soltysh is there any action on this? I'm a bit slow this latest days :)